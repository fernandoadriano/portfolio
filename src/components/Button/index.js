import styled, { css } from 'styled-components';
import propToStyle from 'src/theme/utils/propToStyle';
import { getThemeColor } from 'src/theme';
import typographyVariants from 'src/theme/typographyVariants';
import breakpointsMedia from 'src/theme/utils/breakpointsMedia';

const ButtonGhost = css`
  color: ${({ theme, color }) => getThemeColor(color, theme)};
  background: transparent; 
`;

const ButtonDefault = css`
  color: white;
  background-color: ${({ theme, backgroundColor }) => getThemeColor(backgroundColor, theme)};
  color: ${({ theme, color }) => getThemeColor(color, theme)};
`;

const Button = styled.button`
  border: 0;
  cursor: pointer;
  font-weight: bold;
  opacity: 1;
  
  ${breakpointsMedia({
    xs: css`padding: 12px 12px;`,
    md: css`padding: 12px 26px;`,
  })}  

  ${typographyVariants.smallestException}

  ${(props) => {
    if (props.ghost) {
      return ButtonGhost;
    }
    return ButtonDefault;
  }}

  transition: opacity ${({ theme }) => theme.transition};
  border-radius: ${(props) => props.theme.borderRadius};
  &:hover,
  &:focus {
    opacity: .5;
  }

  ${breakpointsMedia({
    xs: css`
      /* All devices */
      ${typographyVariants.smallestException}
    `,
    md: css`
     /* From md breakpoint */
     ${typographyVariants.buttonText}
    `,
  })}

  &:disabled {
    cursor: not-allowed;
    opacity: .2;
  }

  ${({ fullWidth, size }) => {
    if (fullWidth) {
      return css`
            width: 100%;
        `;
    }
    return css`
            width: ${Math.round((size * 100) / 12)}%;
        `;
  }};

  ${propToStyle('margin')}
  ${propToStyle('display')}
  ${propToStyle('width')}
`;

export default Button;
