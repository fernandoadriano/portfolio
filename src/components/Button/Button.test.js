import React from 'react';
import user from '@testing-library/user-event';
import { render, screen } from 'src/infra/test/testUtils';

import Button from './index';

describe('<Button />', () => {
  describe('Renderização', () => {
    it('simples', () => {
      render(
        <Button>Botão Habilitado</Button>
        ,
      );
      const botao = screen.getByText(/botão habilitado/i);
      expect(botao).not.toHaveAttribute('disabled');
    });

    it('botao desabilitado', () => {
      render(
        <Button
          disabled
        >
          Botão Desabilitado
        </Button>
        ,
      );

      const botao = screen.getByText(/botão desabilitado/i);
      expect(botao).toHaveAttribute('disabled');

      expect(true).toBe(true);
    });

    it('botão Ghost', () => {
      render(
        <Button>Botão Ghost</Button>
        ,
      );
      const botao = screen.getByText(/botão ghost/i);

      expect(botao).toMatchSnapshot();
    });
  });

  describe('Interação', () => {
    it('onClick()', () => {
      const onClick = jest.fn();
      render(
        <Button onClick={onClick}>Botão Padrão</Button>
        ,
      );
      const botao = screen.getByText(/botão padrão/i);
      user.click(botao);
      expect(onClick).toHaveBeenCalled();
    });
  });
});
