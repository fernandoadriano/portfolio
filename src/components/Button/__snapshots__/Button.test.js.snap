// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Renderização botão Ghost 1`] = `
.c0 {
  border: 0;
  cursor: pointer;
  font-weight: bold;
  opacity: 1;
  font-size: 12px;
  font-weight: 400;
  line-height: 1;
  color: white;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  border-radius: 15px;
  width: NaN%;
}

.c0:hover,
.c0:focus {
  opacity: .5;
}

.c0:disabled {
  cursor: not-allowed;
  opacity: .2;
}

@media only screen and (min-width:0px) {
  .c0 {
    padding: 12px 12px;
  }
}

@media only screen and (min-width:960px) {
  .c0 {
    padding: 12px 26px;
  }
}

@media only screen and (min-width:0px) {
  .c0 {
    font-size: 12px;
    font-weight: 400;
    line-height: 1;
  }
}

@media only screen and (min-width:960px) {
  .c0 {
    display: block;
    font-family: Cabin;
    font-size: 18px;
    font-weight: 400;
    line-height: 1.3;
    margin-bottom: 10px;
  }
}

<button
  class="c0"
>
  Botão Ghost
</button>
`;
